--- libfreerdp/CMakeLists.txt	2019-07-24 06:11:52.454197500 +0300
+++ ../../informatimago/FreeRDP/libfreerdp/CMakeLists.txt	2019-07-20 18:27:13.142600324 +0300
@@ -69,7 +69,7 @@
 		foreach (_lib ${ARGN})
 						list (APPEND LIBFREERDP_LIBS "${_lib}")
 		endforeach()
-		 set (LIBFREERDP_LIBS ${LIBFREERDP_LIBS} PARENT_SCOPE)
+        set (LIBFREERDP_LIBS ${LIBFREERDP_LIBS} PARENT_SCOPE)
 endmacro()
 
 macro (freerdp_definition_add)
@@ -79,28 +79,6 @@
 		 set (LIBFREERDP_DEFINITIONS ${LIBFREERDP_DEFINITIONS} PARENT_SCOPE)
 endmacro()
 
-if (WITH_SWSCALE)
-    find_package(SWScale REQUIRED)
-endif(WITH_SWSCALE)
-if (WITH_CAIRO)
-    find_package(Cairo REQUIRED)
-endif(WITH_CAIRO)
-
-if (SWScale_FOUND)
-    add_definitions(-DSWSCALE_FOUND=1)
-    include_directories(${SWScale_INCLUDE_DIR})
-    freerdp_library_add(${SWScale_LIBRARY})
-else(SWScale_FOUND)
-
-    if (CAIRO_FOUND)
-        add_definitions(-DCAIRO_FOUND=1)
-        include_directories(${CAIRO_INCLUDE_DIR})
-        freerdp_library_add(${CAIRO_LIBRARY})
-    else(CAIRO_FOUND)
-        message(WARNING "neigter swscale nor libcairo detected, compiling without image scaling support!")
-    endif(CAIRO_FOUND)
-endif(SWScale_FOUND)
-
 set(${MODULE_PREFIX}_SUBMODULES
 	utils
 	common
@@ -114,6 +92,7 @@
 	add_subdirectory(${${MODULE_PREFIX}_SUBMODULE})
 endforeach()
 
+
 ## cmake source properties are only seen by targets in the same CMakeLists.txt
 ## therefore primitives and codecs need to be defined here
 
@@ -228,9 +207,7 @@
 if(WITH_OPENH264)
 	set(CODEC_SRCS ${CODEC_SRCS} codec/h264_openh264.c)
 	freerdp_include_directory_add(${OPENH264_INCLUDE_DIR})
-	if (NOT WITH_OPENH264_LOADING)
-		freerdp_library_add(${OPENH264_LIBRARIES})
-	endif (NOT WITH_OPENH264_LOADING)
+	freerdp_library_add(${OPENH264_LIBRARIES})
 endif()
 
 if(WITH_FFMPEG)
@@ -239,6 +216,28 @@
 	freerdp_library_add(${FFMPEG_LIBRARIES})
 endif()
 
+if(WITH_PKCS11H)
+    freerdp_definition_add(-DWITH_PKCS11H)
+    freerdp_include_directory_add(${PKCS11H_INCLUDE_DIR})
+    freerdp_library_add(${PKCS11H_LIBRARIES})
+    freerdp_include_directory_add(${FREERDP_SCQUERY_INCLUDE_DIR})
+    freerdp_library_add(${FREERDP_SCQUERY_LIBRARIES})
+    add_subdirectory(scquery)
+    if(GSS_FOUND)
+        if (${GSS_FLAVOUR} STREQUAL "MIT")
+            freerdp_definition_add(-DWITH_MIT_KRB5_PKINIT)
+            freerdp_include_directory_add(${FREERDP_MIT_KRB5_PKINIT_INCLUDE_DIR})
+            freerdp_library_add(${GSS_LIBRARIES})
+            freerdp_library_add(${FREERDP_MIT_KRB5_PKINIT_LIBRARIES})
+            add_subdirectory(mit-krb5-pkinit)
+        endif()
+    endif()
+endif()
+
+if(WITH_KERBEROS)
+    freerdp_definition_add(-DWITH_KERBEROS)
+endif()
+
 if(WIN32 AND WITH_MEDIA_FOUNDATION)
 	set(CODEC_SRCS ${CODEC_SRCS} codec/h264_mf.c)
 endif()
@@ -378,7 +377,6 @@
     set_target_properties(${MODULE_NAME} PROPERTIES VERSION ${FREERDP_VERSION} SOVERSION ${FREERDP_API_VERSION})
 endif()
 
-
 target_link_libraries(${MODULE_NAME} ${PRIVATE_KEYWORD} ${LIBFREERDP_LIBS} winpr)
 
 install(TARGETS ${MODULE_NAME} DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT libraries EXPORT FreeRDPTargets)
